File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {number[]} preorder
 * @param {number[]} inorder
 * @return {TreeNode}
 */
var buildTree = function (preorder, inorder) {
    function solve(preorder, inorder, preorderIndex, InorderStart, InorderEnd,n){
        if (InorderStart>InorderEnd ){
            return null;
        }
        const root = new TreeNode(preorder[preorderIndex.index])
        let InorderMiddle;
        for(let i=InorderStart;i<=InorderEnd;i++){
            if (root.val==inorder[i]){
                InorderMiddle=i;
                break;
            }
        }
        preorderIndex.index =  preorderIndex.index +1;
        root.left = solve(preorder, inorder, preorderIndex, InorderStart, InorderMiddle-1,n);
        root.right = solve(preorder, inorder, preorderIndex, InorderMiddle+1, InorderEnd,n);
        return root
    }

    const n = preorder.length;

    let preorderIndex ={index:0}

    let ans= solve(preorder, inorder,preorderIndex, 0,n-1,n)
    return ans

};

File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
        TreeNode* helper(vector<int>& inorder, vector<int>& preorder, int n,
                     int low, int high, int &preorderindex) {
        if (low > high) {
            return NULL;
        }
        int inorderIndex = -1;
        for (int i = low; i <= high; i++) {
            if (inorder[i] == preorder[preorderindex]) {
                inorderIndex = i;
                break;
            }
        }
        TreeNode* root = new TreeNode(inorder[inorderIndex]);
        preorderindex++;
        root->left = helper(inorder, preorder, n, low, inorderIndex - 1,
                            preorderindex);
        root->right = helper(inorder, preorder, n, inorderIndex + 1, high,
                             preorderindex);
        
        
        return root;
    }
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {

         int n = inorder.size();
        int low = 0;
        int high = n - 1;
        int preorderindex = 0;

        return helper(inorder, preorder, n, low, high, preorderindex);
        
    }
};